File: frontend/src/core/derived/asset.ts
import { type Database } from 'sql.js';
import type { Operation } from '../types/operation';
import type { ChangeNotification } from './index';

export function applyAssetOperation(db: Database, op: Operation): ChangeNotification[] {
  const { opType } = op.envelope;
  const payload = op.payload as Record<string, unknown>;
  const nodeId = payload.nodeId as string;

  if (opType === 'node.delete') {
    db.run('DELETE FROM node_asset WHERE node_id = ?', [nodeId]);
    return [{ scope: 'all', nodeId }];
  }

  if (opType === 'asset.upload') {
    db.run(
      `INSERT OR IGNORE INTO node_asset (node_id, asset_hash, mime_type, size, original_name, uploaded_at)
       VALUES (?, ?, ?, ?, ?, ?)`,
      [
        nodeId,
        payload.assetHash as string,
        payload.mimeType as string,
        payload.sizeBytes as number,
        payload.originalName as string,
        new Date().toISOString(),
      ]
    );
    return [{ scope: 'node', nodeId }];
  }

  if (opType === 'asset.delete') {
    const assetHash = payload.assetHash as string | undefined;
    if (assetHash) {
      db.run('DELETE FROM node_asset WHERE node_id = ? AND asset_hash = ?', [nodeId, assetHash]);
    } else {
      db.run('DELETE FROM node_asset WHERE node_id = ?', [nodeId]);
    }
    return [{ scope: 'node', nodeId }];
  }

  return [];
}

File: app/core/derived/asset.py
"""Asset derived-state applier.

Maps a node to the metadata of a content-addressed blob. The blob bytes live
outside the operation log; the derived table is used to resolve a node to its
file metadata for downloads and share rendering.
"""

from __future__ import annotations

import sqlite3

from app.core.operation import Operation


def apply_asset_upload(conn: sqlite3.Connection, op: Operation) -> None:
    """Apply an ``asset.upload`` operation.

    Payload fields:
        - nodeId: UUIDv7 of the asset node.
        - assetHash: SHA-256 of the blob.
        - mimeType: MIME type of the uploaded file.
        - sizeBytes: file size in bytes.
        - originalName: original filename.
    """
    payload = op.payload
    node_id = payload["nodeId"]
    uploaded_at = op.envelope.timestamp.isoformat() if op.envelope.timestamp else None
    conn.execute(
        """
        INSERT INTO node_asset (node_id, asset_hash, mime_type, size, original_name, uploaded_at)
        VALUES (?, ?, ?, ?, ?, ?)
        ON CONFLICT(node_id, asset_hash) DO UPDATE SET
            mime_type = excluded.mime_type,
            size = excluded.size,
            original_name = excluded.original_name,
            uploaded_at = excluded.uploaded_at
        """,
        (
            node_id,
            payload["assetHash"],
            payload["mimeType"],
            payload["sizeBytes"],
            payload["originalName"],
            uploaded_at,
        ),
    )


def apply_asset_delete(conn: sqlite3.Connection, op: Operation) -> None:
    """Apply an ``asset.delete`` operation by removing the node_asset row."""
    node_id = op.payload["nodeId"]
    conn.execute("DELETE FROM node_asset WHERE node_id = ?", (node_id,))
